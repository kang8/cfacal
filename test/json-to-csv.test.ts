import {
  csvHeader,
  formatJson,
  getFirstDayOfNextMonth,
  parseCinema,
  sortByPlayTime,
} from '../src/json-to-csv.ts'
import { format as dateFormat } from 'jsr:@std/datetime@0.225'
import { stringify as csvStringify } from 'jsr:@std/csv@1'
import { assertEquals, assertThrows } from 'jsr:@std/assert@1'
import { assertSnapshot } from 'jsr:@std/testing@1/snapshot'
import { Body, CinemaInfo, Movie } from '../src/types.ts'

Deno.test('Get first day of next month', () => {
  assertEquals(
    dateFormat(getFirstDayOfNextMonth(new Date('2023-04-25')), 'yyyy-MM-dd'),
    '2023-05-01',
  )
  assertEquals(
    dateFormat(getFirstDayOfNextMonth(new Date('2023-11-25')), 'yyyy-MM-dd'),
    '2023-12-01',
  )
  assertEquals(
    dateFormat(getFirstDayOfNextMonth(new Date('2023-12-25')), 'yyyy-MM-dd'),
    '2024-01-01',
  )
})

Deno.test(
  'Iterate through all days of the next month',
  async (t) => {
    const firstDay = getFirstDayOfNextMonth(new Date('2023-04-25'))
    const url = new URL('https://kang.test')

    const days: Array<string> = []

    for (
      const nextDay = new Date(firstDay);
      firstDay.getMonth() === nextDay.getMonth();
      nextDay.setDate(nextDay.getDate() + 1)
    ) {
      url.search = new URLSearchParams({
        playTime: dateFormat(nextDay, 'yyyy-MM-dd HH:mm:ss'),
      }).toString()

      days.push(url.toString())
    }

    await assertSnapshot(t, days)
  },
)

Deno.test('Format json', async (t) => {
  const json = JSON.parse(
    await Deno.readTextFile('./test/fixtures/23_05_07.json'),
  ) as Body

  const res = formatJson(json)

  await assertSnapshot(t, res)
})

Deno.test('Parse cinema', () => {
  assertEquals(parseCinema('小西天艺术影院', '1号厅'), '小西天')
  assertEquals(parseCinema('小西天艺术影院', '2号厅'), '小西天(2)')
  assertEquals(parseCinema('百子湾艺术影院', '1号厅'), '百子湾')
  assertEquals(parseCinema('百子湾艺术影院', '2号厅'), '百子湾')

  assertThrows(
    () => {
      parseCinema('西天艺术影院' as CinemaInfo, '2号厅')
    },
    Error,
    'Do not support cinema: [西天艺术影院].',
  )
})

Deno.test('Sort movies by play time', () => {
  const movies: Movie[] = [
    {
      cinima: '小西天',
      endTime: '2023-05-07 20:40:00',
      englishName: 'Ryuichi Sakamoto: CODA',
      name: '坂本龙一：终曲',
      playTime: '2023-05-07 19:00:00',
      year: 2017,
      director: '史蒂芬·野村·斯奇博',
    },
    {
      cinima: '百子湾',
      endTime: '2023-05-07 15:35:00',
      englishName: 'On the Milky Road',
      name: '牛奶配送员的奇幻人生',
      playTime: '2023-05-07 13:30:00',
      year: 2016,
      director: '埃米尔·库斯图里卡 Emir Kusturiča',
    },
    {
      cinima: '百子湾',
      endTime: '2023-05-07 18:09:00',
      englishName: 'Black Cat, White Cat',
      name: '黑猫白猫',
      playTime: '2023-05-07 16:00:00',
      year: 1998,
      director: '埃米尔·库斯图里卡 Emir Kusturiča',
    },
  ]

  sortByPlayTime(movies)

  assertEquals(movies[0].name, '牛奶配送员的奇幻人生')
  assertEquals(movies[1].name, '黑猫白猫')
  assertEquals(movies[2].name, '坂本龙一：终曲')
})

Deno.test('convert movies to csv', async (t) => {
  const movies: Movie[] = [
    {
      cinima: '小西天',
      endTime: '2023-05-07 20:40:00',
      englishName: 'Ryuichi Sakamoto: CODA',
      name: '坂本龙一：终曲',
      playTime: '2023-05-07 19:00:00',
      year: 2017,
      director: '史蒂芬·野村·斯奇博',
    },
    {
      cinima: '百子湾',
      endTime: '2023-05-07 15:35:00',
      englishName: 'On the Milky Road',
      name: '牛奶配送员的奇幻人生',
      playTime: '2023-05-07 13:30:00',
      year: 2016,
      director: '埃米尔·库斯图里卡 Emir Kusturiča',
    },
    {
      cinima: '百子湾',
      endTime: '2023-05-07 18:09:00',
      englishName: 'Black Cat, White Cat',
      name: '黑猫白猫',
      playTime: '2023-05-07 16:00:00',
      year: 1998,
      director: '埃米尔·库斯图里卡 Emir Kusturiča',
    },
  ]

  const csvStr = csvStringify(movies, {
    columns: csvHeader,
  })

  await assertSnapshot(t, csvStr)
})
