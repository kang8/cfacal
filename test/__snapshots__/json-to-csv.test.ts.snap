export const snapshot = {};

snapshot[`Iterate through all days of the next month 1`] = `
[
  "https://kang.test/?playTime=2023-05-01+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-02+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-03+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-04+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-05+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-06+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-07+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-08+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-09+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-10+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-11+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-12+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-13+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-14+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-15+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-16+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-17+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-18+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-19+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-20+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-21+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-22+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-23+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-24+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-25+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-26+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-27+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-28+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-29+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-30+00%3A00%3A00",
  "https://kang.test/?playTime=2023-05-31+00%3A00%3A00",
]
`;

snapshot[`Format json 1`] = `
[
  {
    cinima: "小西天",
    director: "",
    endTime: "2023-05-07 13:30:00",
    englishName: "What Do We See When We Look at the Sky?",
    name: "当我们仰望天空时看见什么？",
    playTime: "2023-05-07 11:00:00",
    year: 2021,
  },
  {
    cinima: "小西天",
    director: "",
    endTime: "2023-05-07 16:02:00",
    englishName: "Proxima",
    name: "比邻星",
    playTime: "2023-05-07 14:15:00",
    year: 2019,
  },
  {
    cinima: "小西天(2)",
    director: "",
    endTime: "2023-05-07 18:05:00",
    englishName: "Ryuichi Sakamoto: Async at The Park Avenue Armory",
    name: "坂本龙一：异步",
    playTime: "2023-05-07 17:00:00",
    year: 2018,
  },
  {
    cinima: "小西天",
    director: "",
    endTime: "2023-05-07 20:40:00",
    englishName: "Ryuichi Sakamoto: CODA",
    name: "坂本龙一：终曲",
    playTime: "2023-05-07 19:00:00",
    year: 2017,
  },
  {
    cinima: "百子湾",
    director: "",
    endTime: "2023-05-07 15:35:00",
    englishName: "On the Milky Road",
    name: "牛奶配送员的奇幻人生",
    playTime: "2023-05-07 13:30:00",
    year: 2016,
  },
  {
    cinima: "百子湾",
    director: "",
    endTime: "2023-05-07 18:09:00",
    englishName: "Black Cat, White Cat",
    name: "黑猫白猫",
    playTime: "2023-05-07 16:00:00",
    year: 1998,
  },
  {
    cinima: "百子湾",
    director: "",
    endTime: "2023-05-07 21:20:00",
    englishName: "Underground",
    name: "地下",
    playTime: "2023-05-07 18:30:00",
    year: 1995,
  },
]
`;

snapshot[`convert movies to csv 1`] = `
'name,englishName,year,director,cinima,playTime,endTime\\r
坂本龙一：终曲,Ryuichi Sakamoto: CODA,2017,史蒂芬·野村·斯奇博,小西天,2023-05-07 19:00:00,2023-05-07 20:40:00\\r
牛奶配送员的奇幻人生,On the Milky Road,2016,埃米尔·库斯图里卡 Emir Kusturiča,百子湾,2023-05-07 13:30:00,2023-05-07 15:35:00\\r
黑猫白猫,"Black Cat, White Cat",1998,埃米尔·库斯图里卡 Emir Kusturiča,百子湾,2023-05-07 16:00:00,2023-05-07 18:09:00\\r
'
`;
